py/ringbuf.c

```diff
@@ -27,19 +27,31 @@
 
 #include "ringbuf.h"
 
+bool ringbuf_init(ringbuf_t *r, uint8_t *buf, size_t capacity) {
+    r->heap = false;
+    r->buf = buf;
+    r->size = capacity + 1;
```

Comment:
```suggestion
    r->size = capacity;
```

This will leave make the test pass as-is. I think adding the +1 will cause writing outside the buffer too.

Reply:
It might be misleading, but the `size` stores indeed the `capacity+1` by design.
I just followed this rule when added `ringbuf_init`.
Please take a look on the code of following functions:

```
size_t ringbuf_capacity(ringbuf_t *r) {
    return r->size - 1;
}
```

```
bool ringbuf_alloc(ringbuf_t *r, size_t capacity, bool long_lived) {
    r->buf = gc_alloc(capacity + 1, false, long_lived);
    r->size = capacity + 1;
    r->iget = r->iput = 0;
    return r->buf != NULL;
}
```

Reply:
I don't think so. Size is buffer size. Capacity is the number of entries that can be held at once. You have to leave one entry empty so that you can distinguish between an empty ringbuf where `iput == iget` and a full ringbuff where `iput == iget - 1`.

So, `r->size` must match the actual buffer size. Otherwise, you'll write one byte past the end. The `init` arg really should be `size` rather than `capacity` because the buffer has already been allocated. `alloc` can take `capacity` because it does the allocation.

Reply:
@tannewt Thanks for quick response!

Yes, you are right! I overlooked that `ringbuf_alloc` allocates one byte more to store `capacity` number of bytes.